src/main/scala/is/hail/variant/GATKGenotypeView.scala

```diff
@@ -0,0 +1,265 @@
+    }
+    if (idx < 0 || idx >= adArrayLength)
+      throw new ArrayIndexOutOfBoundsException(idx)
+    assert(!m.loadBit(adArrayStart + 4, idx))
+    m.loadInt(adElemsStart + idx * 4)
+  }
+
+  def getDP: Int = m.loadInt(currentElementStart + dpOffset)
+
+  def getGQ: Int = m.loadInt(currentElementStart + gqOffset)
+
+  def getPL(idx: Int): Int = {
+    if (plElemsStart == -1) {
+      plArrayStart = m.loadInt(currentElementStart + pxOffset)
+      plArrayLength = m.loadInt(plArrayStart)
+      plElemsStart = plArrayStart + StructGenotypeView.expectedPL.elementsOffset(plArrayStart)
+    }
+    if (idx < 0 || idx >= plArrayLength)
+      throw new ArrayIndexOutOfBoundsException(idx)
+    assert(!m.loadBit(plArrayStart + 4, idx))
+    m.loadInt(plElemsStart + idx * 4)
+  }
+
+}
+
+object StructGenotypeView {
+  val expectedGT = TCall
+  val expectedAD = TArray(TInt32)
+  val expectedDP = TInt32
+  val expectedGQ = TInt32
+  val expectedPL = TArray(TInt32)
+}
+
+private class StructGenotypeView(rs: TStruct) extends GATKGenotypeView {
+  private val ta = rs.fields(2).typ.asInstanceOf[TArray]
+  private val t = ta.elementType.asInstanceOf[TStruct]
+  private val gsOffset = rs.byteOffsets(2)
+  private val tSize = ta.elementByteSize
+
+  private val hasFieldGT: Boolean = t.fieldIdx.contains("GT") && t.field("GT").typ == StructGenotypeView.expectedGT
+  private val hasFieldAD: Boolean = t.fieldIdx.contains("AD") && t.field("AD").typ == StructGenotypeView.expectedAD
+  private val hasFieldDP: Boolean = t.fieldIdx.contains("DP") && t.field("DP").typ == StructGenotypeView.expectedDP
+  private val hasFieldGQ: Boolean = t.fieldIdx.contains("GQ") && t.field("GQ").typ == StructGenotypeView.expectedGQ
+  private val hasFieldPL: Boolean = t.fieldIdx.contains("PL") && t.field("PL").typ == StructGenotypeView.expectedPL
+
+  private val (gtIndex, gtOffset) = if (hasFieldGT) {
+    val gtIndex = t.fieldIdx("GT")
+    val gtOffset = t.byteOffsets(gtIndex)
+    (gtIndex, gtOffset)
+  } else (uninitialized[Int], uninitialized[Int])
+
+  private val (adIndex, adOffset) = if (hasFieldAD) {
+    val adIndex = t.fieldIdx("AD")
+    val adOffset = t.byteOffsets(adIndex)
+    (adIndex, adOffset)
+  } else (uninitialized[Int], uninitialized[Int])
+
+  private val (dpIndex, dpOffset) = if (hasFieldDP) {
+    val dpIndex = t.fieldIdx("DP")
+    val dpOffset = t.byteOffsets(dpIndex)
+    (dpIndex, dpOffset)
+  } else (uninitialized[Int], uninitialized[Int])
+
+  private val (gqIndex, gqOffset) = if (hasFieldGQ) {
+    val gqIndex = t.fieldIdx("GQ")
+    val gqOffset = t.byteOffsets(gqIndex)
+    (gqIndex, gqOffset)
+  } else (uninitialized[Int], uninitialized[Int])
+
+  private val (plIndex, plOffset) = if (hasFieldPL) {
+    val plIndex = t.fieldIdx("PL")
+    val plOffset = t.byteOffsets(plIndex)
+    (plIndex, plOffset)
+  } else (uninitialized[Int], uninitialized[Int])
+
+  private var m: MemoryBuffer = _
+  private var arrStart: Int = _
+  private var arrSize: Int = _
```

Comment:
I'd say `gsLength`.  Size is bytes.  I think we've been using this consistently.